File: src/double_ok_gesture/recognizer.py
"""Runtime recognizer for two-hand OK gestures."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Iterable

import numpy as np

from .features import as_landmark_array, feature_names, feature_vector, rule_ok_score
from .model_io import load_model_artifact


@dataclass
class HandPrediction:
    handedness: str
    ok_score: float
    is_ok: bool
    landmarks: np.ndarray


@dataclass
class DoubleOKResult:
    hands: list[HandPrediction]
    double_ok: bool
    stable_double_ok: bool
    ok_count: int


class OKHandClassifier:
    """OK vs not-OK classifier with rule fallback."""

    def __init__(self, model_path: str | None = None, threshold: float = 0.68) -> None:
        if not 0.0 <= threshold <= 1.0:
            raise ValueError("threshold must be in [0.0, 1.0]")

        self.model = None
        if model_path is not None:
            artifact = load_model_artifact(model_path)
            expected_features = artifact.metadata.get("feature_columns")
            if expected_features and list(expected_features) != feature_names():
                raise ValueError("Model feature schema does not match the runtime feature schema")
            self.model = artifact.model
        self.threshold = threshold

    def score(self, landmarks: Iterable[object] | np.ndarray, handedness: str | None = None) -> float:
        if self.model is None:
            return rule_ok_score(landmarks, handedness)

        vector = feature_vector(landmarks, handedness).reshape(1, -1)
        if hasattr(self.model, "predict_proba"):
            proba = self.model.predict_proba(vector)[0]
            classes = list(getattr(self.model, "classes_", [0, 1]))
            if 1 in classes:
                score = float(proba[classes.index(1)])
            else:
                score = float(proba[-1])
            return _bounded_score(score)
        if hasattr(self.model, "decision_function"):
            raw = float(self.model.decision_function(vector)[0])
            return _bounded_score(float(1.0 / (1.0 + np.exp(-raw))))
        return _bounded_score(float(self.model.predict(vector)[0]))

    def predict(
        self,
        landmarks: Iterable[object] | np.ndarray,
        handedness: str | None = None,
    ) -> HandPrediction:
        points = as_landmark_array(landmarks)
        score = self.score(points, handedness)
        return HandPrediction(
            handedness=handedness or "Unknown",
            ok_score=score,
            is_ok=score >= self.threshold,
            landmarks=points,
        )


class DoubleOKRecognizer:
    """Detect hands, classify each hand, and decide whether both are OK."""

    def __init__(
        self,
        model_path: str | None = None,
        ok_threshold: float = 0.68,
        max_num_hands: int = 2,
        stable_window: int = 5,
        stable_min_positive: int = 3,
        min_detection_confidence: float = 0.55,
        min_tracking_confidence: float = 0.55,
    ) -> None:
        if max_num_hands < 2:
            raise ValueError("max_num_hands must be at least 2")
        if stable_window < 1:
            raise ValueError("stable_window must be at least 1")
        if not 1 <= stable_min_positive <= stable_window:
            raise ValueError("stable_min_positive must be between 1 and stable_window")
        for name, value in (
            ("min_detection_confidence", min_detection_confidence),
            ("min_tracking_confidence", min_tracking_confidence),
        ):
            if not 0.0 <= value <= 1.0:
                raise ValueError(f"{name} must be in [0.0, 1.0]")

        self.classifier = OKHandClassifier(model_path=model_path, threshold=ok_threshold)
        self.stable_min_positive = stable_min_positive
        self.history: deque[bool] = deque(maxlen=stable_window)

        import mediapipe as mp

        self._mp = mp
        self._hands = mp.solutions.hands.Hands(
            static_image_mode=False,
            max_num_hands=max_num_hands,
            min_detection_confidence=min_detection_confidence,
            min_tracking_confidence=min_tracking_confidence,
        )

    def close(self) -> None:
        self._hands.close()

    def __enter__(self) -> "DoubleOKRecognizer":
        return self

    def __exit__(self, _exc_type, _exc_value, _traceback) -> None:
        self.close()

    def process_rgb(self, frame_rgb: np.ndarray) -> DoubleOKResult:
        result = self._hands.process(frame_rgb)
        predictions: list[HandPrediction] = []

        if result.multi_hand_landmarks:
            handedness_labels = []
            if result.multi_handedness:
                for item in result.multi_handedness:
                    handedness_labels.append(item.classification[0].label)

            for index, hand_landmarks in enumerate(result.multi_hand_landmarks):
                handedness = handedness_labels[index] if index < len(handedness_labels) else "Unknown"
                points = np.asarray(
                    [[lm.x, lm.y, lm.z] for lm in hand_landmarks.landmark],
                    dtype=np.float32,
                )
                predictions.append(self.classifier.predict(points, handedness))

        ok_count = sum(1 for hand in predictions if hand.is_ok)
        double_ok = ok_count >= 2
        self.history.append(double_ok)
        stable_double_ok = sum(self.history) >= self.stable_min_positive

        return DoubleOKResult(
            hands=predictions,
            double_ok=double_ok,
            stable_double_ok=stable_double_ok,
            ok_count=ok_count,
        )

    def process_bgr(self, frame_bgr: np.ndarray) -> DoubleOKResult:
        import cv2

        frame_rgb = cv2.cvtColor(frame_bgr, cv2.COLOR_BGR2RGB)
        return self.process_rgb(frame_rgb)

    def draw(self, frame_bgr: np.ndarray, result: DoubleOKResult) -> np.ndarray:
        from .live_ui import draw_hand_tracking

        return draw_hand_tracking(frame_bgr, result)


def _bounded_score(score: float) -> float:
    if not np.isfinite(score):
        raise ValueError("Classifier produced a non-finite score")
    return float(np.clip(score, 0.0, 1.0))

File: src/recognizer.cpp
#include "double_ok_gesture/recognizer.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace double_ok_gesture {
namespace {

void validate_threshold(double threshold) {
    if (!std::isfinite(threshold) || threshold < 0.0 || threshold > 1.0) {
        throw std::invalid_argument("threshold must be finite and in [0.0, 1.0]");
    }
}

}  // namespace

OKHandClassifier::OKHandClassifier(double threshold) : threshold_(threshold) {
    validate_threshold(threshold_);
}

double OKHandClassifier::score(const Landmarks& landmarks, const std::string& handedness) const {
    return rule_ok_score(landmarks, handedness);
}

HandPrediction OKHandClassifier::predict(const Landmarks& landmarks, const std::string& handedness) const {
    const double ok = score(landmarks, handedness);
    return predict_with_score(landmarks, handedness, ok, false);
}

HandPrediction OKHandClassifier::predict_with_score(
    const Landmarks& landmarks,
    const std::string& handedness,
    double ok_score,
    bool landmarks_estimated) const {
    const double ok = bounded_score(ok_score);
    return {
        handedness.empty() ? "Unknown" : handedness,
        ok,
        ok >= threshold_,
        landmarks,
        landmarks_estimated,
        std::nullopt,
        std::nullopt,
        0.0,
    };
}

double OKHandClassifier::threshold() const {
    return threshold_;
}

DoubleOKRecognizer::DoubleOKRecognizer(
    OKHandClassifier classifier,
    std::size_t stable_window,
    std::size_t stable_min_positive)
    : classifier_(std::move(classifier)), stable_min_positive_(stable_min_positive),
      history_limit_(stable_window) {
    if (stable_window < 1) {
        throw std::invalid_argument("stable_window must be at least 1");
    }
    if (stable_min_positive < 1 || stable_min_positive > stable_window) {
        throw std::invalid_argument("stable_min_positive must be between 1 and stable_window");
    }
}

DoubleOKResult DoubleOKRecognizer::process_hands(const std::vector<DetectedHand>& hands) {
    DoubleOKResult result;
    result.hands.reserve(hands.size());
    for (const DetectedHand& hand : hands) {
        if (hand.ok_score) {
            HandPrediction prediction = classifier_.predict_with_score(
                hand.landmarks,
                hand.handedness,
                *hand.ok_score,
                hand.landmarks_estimated);
            prediction.landmark_confidences = hand.landmark_confidences;
            prediction.box = hand.box;
            prediction.handedness_confidence =
                hand.handedness_confidence;
            result.hands.push_back(std::move(prediction));
        } else {
            const Landmarks& classifier_landmarks =
                hand.metric_landmarks
                    ? *hand.metric_landmarks
                    : hand.landmarks;
            HandPrediction prediction =
                !hand.gesture_landmarks_reliable
                    ? classifier_.predict_with_score(
                          classifier_landmarks,
                          hand.handedness,
                          0.0,
                          hand.landmarks_estimated)
                    : classifier_.predict(
                          classifier_landmarks, hand.handedness);
            // metric_landmarks are classification-only. Preserve normalized
            // coordinates for overlays and capture-gate geometry.
            prediction.landmarks = hand.landmarks;
            prediction.landmarks_estimated = hand.landmarks_estimated;
            prediction.landmark_confidences = hand.landmark_confidences;
            prediction.box = hand.box;
            prediction.handedness_confidence = hand.handedness_confidence;
            result.hands.push_back(prediction);
        }
    }
    result.ok_count = static_cast<int>(std::count_if(result.hands.begin(), result.hands.end(), [](const auto& hand) {
        return hand.is_ok;
    }));
    result.double_ok = result.ok_count >= 2;

    history_.push_back(result.double_ok);
    while (history_.size() > history_limit_) {
        history_.pop_front();
    }
    const int positives = static_cast<int>(std::count(history_.begin(), history_.end(), true));
    result.stable_double_ok = positives >= static_cast<int>(stable_min_positive_);
    return result;
}

void DoubleOKRecognizer::reset() {
    history_.clear();
}

double bounded_score(double score) {
    if (!std::isfinite(score)) {
        throw std::runtime_error("Classifier produced a non-finite score");
    }
    return std::clamp(score, 0.0, 1.0);
}

}  // namespace double_ok_gesture
